File: core/src/main/java/com/mcreater/amclcore/concurrent/task/AbstractTask.java
package com.mcreater.amclcore.concurrent.task;

import com.mcreater.amclcore.concurrent.TaskState;
import com.mcreater.amclcore.exceptions.report.ExceptionReporter;
import com.mcreater.amclcore.i18n.Text;
import com.mcreater.amclcore.util.sets.ImmutableDoubleValueSet;
import lombok.Getter;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.Nullable;

import javax.naming.OperationNotSupportedException;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.Vector;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.RecursiveTask;
import java.util.function.Consumer;

import static com.mcreater.amclcore.concurrent.ConcurrentExecutors.INTERFACE_EVENT_EXECUTORS;
import static com.mcreater.amclcore.concurrent.ConcurrentExecutors.createInterfaceEventExecutor;
import static com.mcreater.amclcore.i18n.I18NManager.translatable;

public abstract class AbstractTask<T> extends RecursiveTask<Optional<T>> {
    public static void printTextData(TaskState<?> taskState, Consumer<String> output) {
        Optional.ofNullable(taskState)
                .map(TaskState::getMessage)
                .map(Text::getText)
                .ifPresent(output);
    }

    private static final Logger EVENT_LOGGER = LogManager.getLogger(AbstractTask.class);
    private final List<Consumer<TaskState<T>>> stateConsumers = new Vector<>();
    private final List<Consumer<AbstractTask<?>>> bindConsumers = new Vector<>();
    @Getter
    private TaskState<T> state;
    private final List<AbstractTask<?>> bindTasks = new Vector<>();
    @Getter
    private AbstractTask<?> topTask;

    public AbstractTask<T> addStateConsumers(List<Consumer<TaskState<T>>> c) {
        c.forEach(this::addStateConsumer);
        return this;
    }

    public AbstractTask<T> addStateConsumer(Consumer<TaskState<T>> c) {
        stateConsumers.add(c);
        return this;
    }

    public AbstractTask<T> addBindConsumers(List<Consumer<AbstractTask<?>>> c) {
        c.forEach(this::addBindConsumer);
        return this;
    }

    public AbstractTask<T> addBindConsumer(Consumer<AbstractTask<?>> c) {
        bindConsumers.add(c);
        return this;
    }

    public List<Consumer<TaskState<T>>> getStateConsumers() {
        return Collections.unmodifiableList(stateConsumers);
    }

    public List<Consumer<AbstractTask<?>>> getBindConsumers() {
        return Collections.unmodifiableList(bindConsumers);
    }

    public AbstractTask() {
        super();
        INTERFACE_EVENT_EXECUTORS.put(this, createInterfaceEventExecutor());
    }

    protected void setState(TaskState<T> state) {
        this.state = state;
        INTERFACE_EVENT_EXECUTORS.get(this).execute(() -> getStateConsumers().forEach(c -> c.accept(state)));
    }

    /**
     * abstract task for implement<br>
     * 用于实现的抽象任务
     *
     * @return the task result<br>任务结果
     * @throws Exception when task throws an exception<br>当任务抛出一个异常
     */
    protected abstract T call() throws Exception;

    private void stateFinish(ImmutableDoubleValueSet<Integer, Integer> lastState, T result) {
        setState(TaskState.<T>builder()
                .taskType(TaskState.Type.FINISHED)
                .totalStage(lastState.getValue1())
                .currentStage(lastState.getValue1())
                .result(result)
                .build()
        );
    }

    private void stateExc(ImmutableDoubleValueSet<Integer, Integer> lastState, Throwable e) {
        setState(TaskState.<T>builder()
                .message(translatable("core.concurrent.base.event.exception.text", e))
                .throwable(e)
                .totalStage(lastState.getValue1())
                .currentStage(lastState.getValue2())
                .taskType(TaskState.Type.ERROR)
                .build()
        );
    }

    private ImmutableDoubleValueSet<Integer, Integer> fetchTaskState() {
        return ImmutableDoubleValueSet.<Integer, Integer>builder()
                .value1(Optional.ofNullable(state).map(TaskState::getTotalStage).orElse(1))
                .value2(Optional.ofNullable(state).map(TaskState::getCurrentStage).orElse(1))
                .build();
    }

    protected Optional<T> compute() {
        ImmutableDoubleValueSet<Integer, Integer> lastState = fetchTaskState();
        try {
            T result = call();
            EVENT_LOGGER.info(translatable("core.concurrent.base.event.finish.name", this).getText());
            stateFinish(lastState, result);
            return Optional.ofNullable(result);
        } catch (Exception e) {
            ExceptionReporter.report(e, ExceptionReporter.ExceptionType.CONCURRENT);
            stateExc(lastState, e);
        }
        return Optional.empty();
    }

    private void addSubTask(AbstractTask<?> task) throws OperationNotSupportedException {
        if (task == this) throw new OperationNotSupportedException("task == this!");
        if (task == null) throw new NullPointerException("task == null!");
        bindTasks.add(task);
        task.topTask = this;
        INTERFACE_EVENT_EXECUTORS.get(this).execute(() -> getBindConsumers().forEach(c -> c.accept(task)));
    }

    public AbstractTask<T> bindTo(@Nullable AbstractTask<?> task) {
        fork();
        Optional.ofNullable(task).ifPresent(t -> {
            try {
                t.addSubTask(AbstractTask.this);
            } catch (Exception e) {
                ExceptionReporter.report(e, ExceptionReporter.ExceptionType.UNKNOWN);
            }
        });
        return this;
    }

    public AbstractTask<T> bind() {
        return bindTo(null);
    }

    public List<AbstractTask<?>> getSubTasks() {
        return Collections.unmodifiableList(bindTasks);
    }

    protected abstract Text getTaskName();

    public String toString() {
        return getTaskName().getText();
    }

    public boolean tryUnfork() {
        throw new RuntimeException(new OperationNotSupportedException());
    }

    public void complete(Optional<T> value) {
        super.complete(value);
        stateFinish(fetchTaskState(), value.orElse(null));
    }

    public void completeExceptionally(Throwable ex) {
        super.completeExceptionally(ex);
        stateExc(fetchTaskState(), ex);
    }

    public AbstractTask<T> submitTo(ForkJoinPool pool) {
        pool.execute(this);
        return this;
    }
}

File: core/src/main/java/com/mcreater/amclcore/concurrent/task/AbstractTask.kt
package com.mcreater.amclcore.concurrent.task

import com.mcreater.amclcore.concurrent.ConcurrentExecutors
import com.mcreater.amclcore.concurrent.TaskState
import com.mcreater.amclcore.exceptions.report.ExceptionReporter
import com.mcreater.amclcore.i18n.I18NManager.Companion.translatable
import com.mcreater.amclcore.i18n.Text
import org.apache.commons.lang3.tuple.ImmutablePair
import org.apache.logging.log4j.LogManager
import java.util.*
import java.util.concurrent.ForkJoinPool
import java.util.concurrent.RecursiveTask
import java.util.function.Consumer
import javax.naming.OperationNotSupportedException


abstract class AbstractTask<T : Any?> : RecursiveTask<Optional<T>>() {
    val stateConsumers: MutableList<Consumer<TaskState<T>>> = Vector()
    val bindConsumers: MutableList<Consumer<AbstractTask<out Any?>>> = Vector()

    private var state: TaskState<T>? = null
    private val bindTasks: MutableList<AbstractTask<out Any?>> = Vector()
    protected var topTask: AbstractTask<*>? = null
    fun getState(): TaskState<T>? {
        return state
    }

    fun addStateConsumers(c: List<Consumer<TaskState<T>>>): AbstractTask<T> {
        c.forEach { addStateConsumer(it) }
        return this
    }

    fun addStateConsumer(c: Consumer<TaskState<T>>): AbstractTask<T> {
        stateConsumers.add(c)
        return this
    }

    fun addBindConsumers(c: List<Consumer<AbstractTask<out Any?>>>): AbstractTask<T> {
        c.forEach { addBindConsumer(it) }
        return this
    }

    fun addBindConsumer(c: Consumer<AbstractTask<out Any?>>): AbstractTask<T> {
        bindConsumers.add(c)
        return this
    }

    init {
        ConcurrentExecutors.INTERFACE_EVENT_EXECUTORS[this] = ConcurrentExecutors.createInterfaceEventExecutor()
    }

    protected fun setState(state: TaskState<T>) {
        this.state = state
        Optional.ofNullable(ConcurrentExecutors.INTERFACE_EVENT_EXECUTORS[this]).ifPresent { a: ForkJoinPool ->
            a.execute {
                stateConsumers.forEach(Consumer { c: Consumer<TaskState<T>> ->
                    c.accept(
                        state
                    )
                })
            }
        }
    }

    /**
     * abstract task for implement<br></br>
     * 用于实现的抽象任务
     *
     * @return the task result<br></br>任务结果
     * @throws Exception when task throws an exception<br></br>当任务抛出一个异常
     */
    @Throws(Exception::class)
    protected abstract fun call(): T
    private fun stateFinish(lastState: ImmutablePair<Int, Int>, result: T?) {
        setState(
            TaskState(
                TaskState.Type.FINISHED,
                null,
                result,
                lastState.key, lastState.key, null
            )
        )
    }

    private fun stateExc(lastState: ImmutablePair<Int, Int>, e: Throwable) {
        setState(
            TaskState(
                TaskState.Type.ERROR,
                e,
                null,
                lastState.key, lastState.value, translatable(
                    "core.concurrent.base.event.exception.text", e.javaClass.getName(),
                    getTaskName().text!!
                )
            )
        )
    }

    private fun fetchTaskState(): ImmutablePair<Int, Int> {
        return ImmutablePair(
            Optional.ofNullable(state).map { it.totalStage }.orElse(1),
            Optional.ofNullable(state).map { it.currentStage }.orElse(1)
        )
    }

    override fun compute(): Optional<T> {
        val lastState = fetchTaskState()
        try {
            val result = call()
            if ("" != this.toString()) EVENT_LOGGER.info(
                translatable(
                    "core.concurrent.base.event.finish.name",
                    this
                ).text
            )
            stateFinish(lastState, result)
            return Optional.ofNullable(result) as Optional<T>
        } catch (e: Exception) {
            ExceptionReporter.report(e, ExceptionReporter.ExceptionType.CONCURRENT)
            stateExc(lastState, e)
        }
        return Optional.empty<T>() as Optional<T>
    }

    @Throws(OperationNotSupportedException::class)
    private fun addSubTask(task: AbstractTask<out Any?>) {
        if (task === this) throw OperationNotSupportedException("task == this!")
        bindTasks.add(task)
        task.topTask = this
        ConcurrentExecutors.INTERFACE_EVENT_EXECUTORS[this]!!.execute {
            bindConsumers.forEach { it.accept(task) }
        }
    }

    fun bindTo(task: AbstractTask<*>?): AbstractTask<T> {
        fork()
        Optional.ofNullable(task).ifPresent {
            try {
                it.addSubTask(this@AbstractTask)
            } catch (e: Exception) {
                ExceptionReporter.report(e, ExceptionReporter.ExceptionType.UNKNOWN)
            }
        }
        return this
    }

    fun bind(): AbstractTask<T> {
        return bindTo(null)
    }

    val subTasks: List<AbstractTask<*>>
        get() = Collections.unmodifiableList(bindTasks)

    protected abstract fun getTaskName(): Text

    override fun toString(): String {
        return getTaskName().text!!
    }

    override fun tryUnfork(): Boolean {
        throw RuntimeException(OperationNotSupportedException())
    }

    override fun complete(value: Optional<T>?) {
        super.complete(value)
        stateFinish(fetchTaskState(), value?.orElse(null))
    }

    override fun completeExceptionally(ex: Throwable) {
        super.completeExceptionally(ex)
        stateExc(fetchTaskState(), ex)
    }

    fun submitTo(pool: ForkJoinPool): AbstractTask<T> {
        pool.execute(this)
        return this
    }

    companion object {
        @JvmStatic
        fun printTextData(taskState: TaskState<*>?, output: Consumer<String>?) {
            Optional.ofNullable(taskState)
                .map { it.message }
                .map { it?.text }
                .ifPresent(output!!)
        }

        @JvmStatic
        private val EVENT_LOGGER = LogManager.getLogger(
            AbstractTask::class.java
        )
    }
}